pippy/PipelineDriver.py

```diff
@@ -593,32 +592,63 @@ def extract_value_ref_args(arg):
             for fut in _futures:
                 fut.wait()
 
-        with self.value_store_lock:
+        with self.value_store_cv:
             assert output_unique_key not in self.value_store
             self.value_store[output_unique_key] = RefcountedFuture(future, output_refcount)
+            self.value_store_cv.notify_all()
 
-        return ValueReference(self.stage_id, output_unique_key)
+        finish_ts = time.time()
+        self.record_event(rank=self.rank_worker.rank, start_ts=start_ts, finish_ts=finish_ts, id=id, name=name, type='received', mbid=cur_microbatch)
+        self.record_event_dependency(from_id=name, to_id=target_name, type='waiting')
 
-    def index_value(self, output_unique_key : str, output_refcount : int, value_ref, idx):
-        # For the purposes of refcounting, decrement this use
-        with self.value_store_lock:
-            refcounted_future = self.value_store[value_ref.unique_key]
-            if refcounted_future.release():
-                self.value_store.pop(value_ref.unique_key)
+        return ValueReference(self.stage_id, output_unique_key)
 
-            indexed = refcounted_future.future.then(lambda f: f.value()[idx])
+    def coalesce_index_value(self, indices : List[Tuple]):
```

Comment:
nit: `coalesce` sounds like the verb here, but i think that should be an adjective, like "coalesced index value". Maybe change to `coalesced`

Reply:
nit: also maybe have a more precise type annotation for `indices` by typing the contained values, e.g. `List[Tuple[str, int, ValueRef, int]]` (not sure if that's right but you get the idea)

Reply:
Thanks, fixed.